### PR TITLE
fix(CDAP-19802): form primary key by correct order

### PR DESCRIPTION
**Pull request by de-lan**
The change:

When we form the table schema directly from DB, we form primary keys by the order of fields. This wouldn't be a problem if the StructuredTableSpecification is defined with primary keys and fields having the same order.  But in the case that they have different orders, we need to correctly honor the order of primary keys' indexes

### CODE REVIEW

File: cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTableAdmin.java [hunk 1]
```diff
@@ -210,42 +208,23 @@ private void addColumns(Connection connection, StructuredTableSchema schema)
...
-      throw new IOException(String.format("Error adding indices to table: %s", schema.getTableId().getName()), e);
+      throw new IOException(String.format("Error adding indices to table: %s", newSchema.getTableId().getName()), e);
     }
   }
 
-  private Set<String> getNonExistIndexes(Connection connection,
-                                         StructuredTableSchema schema) throws SQLException {
-    DatabaseMetaData metaData = connection.getMetaData();
-    Set<String> existingIndexes = new HashSet<>();
-    try (ResultSet rs = metaData.getIndexInfo(null, null,
```
Comment (de-lan) [inline cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTableAdmin.java:234]: Remove unnecessary DB queries to construct columns and indexes, since we have them loaded in cache

File: cdap-storage-ext-spanner/src/main/java/io/cdap/cdap/storage/spanner/SpannerStructuredTableAdmin.java [hunk 2]
```diff
@@ -230,7 +233,15 @@ private StructuredTableSchema loadSchema(StructuredTableId tableId) throws Table
       throw new TableNotFoundException(tableId);
     }
 
-    return new StructuredTableSchema(tableId, fields, primaryKeys, indexes);
+    List<String> primaryKeys = primaryKeysOrderMap.keySet()
+      .stream()
+      .sorted()
```
Comment (chtyim) [inline cdap-storage-ext-spanner/src/main/java/io/cdap/cdap/storage/spanner/SpannerStructuredTableAdmin.java]: What the sorting is for? If you need the keys sorted, can we use a SortedMap instead?

Comment (de-lan) [reply cdap-storage-ext-spanner/src/main/java/io/cdap/cdap/storage/spanner/SpannerStructuredTableAdmin.java]: updated with TreeMap

**Review by chtyim — APPROVED** [review]

File: cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTableAdmin.java [hunk 3]
```diff
@@ -151,8 +152,7 @@ private void createTable(StructuredTableSpecification spec) throws IOException {
         }
 
         // Create indexes
-        Set<String> nonExistIndexes = getNonExistIndexes(connection, new StructuredTableSchema(spec));
-        for (String indexStatement : getCreateIndexStatements(spec.getTableId(), nonExistIndexes)) {
+        for (String indexStatement : getCreateIndexStatements(spec.getTableId(), new HashSet<>(spec.getIndexes()))) {
```
Comment (tivv) [inline cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTableAdmin.java:157]: Why this change? I mean, it would work, but you would issue much more "Create index" statements.

File: cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTableAdmin.java [hunk 4]
```diff
@@ -186,14 +186,13 @@ private boolean tableExistsInternal(Connection connection, StructuredTableId tab
   }
 
   // Add new non-primary key columns to table
-  private void addColumns(Connection connection, StructuredTableSchema schema)
+  private void addColumns(Connection connection, StructuredTableSchema newSchema, StructuredTableSchema oldSchema)
```
Comment (tivv) [inline cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTableAdmin.java]: nit: use `existingSchema` to be consistent with other methods.

File: cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTableAdmin.java [hunk 5]
```diff
@@ -371,13 +342,19 @@ private StructuredTableSchema loadSchema(StructuredTableId tableId)
...
+        // "t" as true indicates primary key index
+        // "f" as false indicates normal index, but it can still be a primary key field that is part of other index
+        // null means it's just a column
+        if ("f".equalsIgnoreCase(indexType)) {
           indexes.add(columnName);
         }
       }
+
+      // Calling MetaData.getPrimaryKeys() to construct table primary keys and maintain the order
+      ResultSet rs = connection.getMetaData().getPrimaryKeys(null, null, tableId.getName());
```
Comment (tivv) [inline cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTableAdmin.java]: I believe the whole idea of the schemaStatement above was to get all data in one query. That query can be enhanced to add PK field order. No need to do another SQL

**Review by tivv — CHANGES_REQUESTED** [review]

File: cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTableAdmin.java (hunk 3 above)
Comment (de-lan) [reply cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTableAdmin.java:157]: This is creating table senario, the table does not exist yet

Comment (de-lan) [reply cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTableAdmin.java:157]: Okay there can be edge case that DB is being created at the same time by two instances... ~~will revert back using DB query result~~
Even in this edge case, we're using `CREATE INDEX IF NOT EXISTS`, it should be safe.
 
And we have this change because to it would be not necessary to query db for index info, we already have it

Comment (de-lan) [reply cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTableAdmin.java:157]: ref: https://github.com/cdapio/cdap/pull/14670/files#r1001251902

Comment (tivv) [reply cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTableAdmin.java:157]: I mean, it's trading one query (select) for potentially multiple (create index). It's probably fine as it's corner case and it's safe anyway

File: cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTableAdmin.java (hunk 5 above)
Comment (de-lan) [reply cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTableAdmin.java]: That is true. However the `ordinal_position` in information_schema is by columns at DB creation time. One query is possible but need more complexed query statement. More investigation is needed on it

And it's not too bad to have one more query while loading schema from DB: 1. It's PK, it's currently immutable in CDAP system; 2. It's one time thing when the instance launch

File: cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTableAdmin.java (hunk 4 above)
Comment (de-lan) [reply cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTableAdmin.java]: updated, pls review

File: cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTableAdmin.java (hunk 5 above)
Comment (tivv) [reply cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTableAdmin.java]: There is no "instance launch" thing, I believe it's something that runs on every pod start, so it's actually pretty important

Comment (tivv) [reply cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTableAdmin.java]: I believe the query change would be relatively simple. You would need to add something like `array_position(i.indkey, a.attnum)` and return it as a separate column (say `index_order`). Then you use it to resort on java site

**Review by tivv — CHANGES_REQUESTED** [review]
See answer inline

File: cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTableAdmin.java (hunk 5 above)
Comment (de-lan) [reply cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTableAdmin.java]: brilliant! It works. I've pushed the change

**Review by tivv — APPROVED** [review]
